### PR TITLE
Containerized integration test environment

### PR DESCRIPTION
**Pull request by ahmetalpbalkan**
Built a containerized environment that looks like an actual VM
and wrote a fake agent which you can invoke the extension binary
through.

Refer to the integration-test/env/README.md for more info.

This helps us test:
- agent binary can handle the directory structure presented by waagent
- agent binary can handle the HandlerEnvironment, log dirs etc
- agent binary decrypts/parses the settings correctly
- agent binary can write status to the right place

By containerizing there will be no residue left on the dev machine
while testing the changes.

### CODE REVIEW

File: integration-test/env/D6608DD84293F393E2206151B84A4426A5C04B06.prv [hunk 1]
```diff
@@ -0,0 +1,33 @@
+Bag Attributes
+    localKeyID: 01 00 00 00
+    Microsoft CSP Name: Microsoft Enhanced Cryptographic Provider v1.0
+Key Attributes
+    X509v3 Key Usage: 10
+-----BEGIN PRIVATE KEY-----
```
Comment (paulmey) [inline integration-test/env/D6608DD84293F393E2206151B84A4426A5C04B06.prv]: Not really recommended to check in private keys, even if they are for tests. Consider writing a script that generates a keypair and encodes the relevant data for the tests.

File: integration-test/env/Extension/bin/_placeholder [hunk 2]
```diff
@@ -0,0 +1 @@
+placeholder for git to preserve this directory
```
Comment (paulmey) [inline integration-test/env/Extension/bin/_placeholder]: Seems unnecessary? the COPY command in the docker file creates this directory if it doesn't exist...

File: test.Dockerfile [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+
+# Copy the test environment
+WORKDIR /var/lib/waagent
+COPY integration-test/env/ .
+RUN ln -s /var/lib/waagent/fake-waagent /sbin/fake-waagent
+
+# Copy the handler files
+COPY HandlerManifest.json ./Extension/
+COPY bin/custom-script-extension ./Extension/bin/
```
Comment (paulmey) [inline test.Dockerfile]: I assume that bin/custom-script-extension is created by #29

File: integration-test/env/Extension/bin/_placeholder (hunk 2 above)
Comment (ahmetalpbalkan) [inline integration-test/env/Extension/bin/_placeholder]: @paulmey but the directory does not appear in .git at all, for documentation purposes (see README.md) I thought this dir better be present when the repo is checked out?

**Comment by paulmey** [comment]
I would not check in any private key. LGTM otherwise

File: test.Dockerfile (hunk 3 above)
Comment (ahmetalpbalkan) [inline test.Dockerfile]: True.

File: integration-test/env/D6608DD84293F393E2206151B84A4426A5C04B06.prv (hunk 1 above)
Comment (paulmey) [inline integration-test/env/D6608DD84293F393E2206151B84A4426A5C04B06.prv]: Generate:
```
openssl req -x509 -newkey rsa:2048 -keyout key.pem -out cert.pem -days 365 -nodes -batch
export FP=$(openssl x509 -in cert.pem -fingerprint -noout|sed  's/.*=\|://g')
mv cert.pem $FP.crt
mv key.pem $FP.prv
```
Encode: (put stdout in protected settings)
```
openssl smime -inkey $FP.prv -encrypt -outform DER $FP.crt <script |base64 -w0
```

Comment (ahmetalpbalkan) [inline integration-test/env/D6608DD84293F393E2206151B84A4426A5C04B06.prv]: damn you just made it look easy. I guess I will make an utility for this.

Comment (paulmey) [inline integration-test/env/D6608DD84293F393E2206151B84A4426A5C04B06.prv]: 👍

File: integration-test/env/Extension/bin/_placeholder (hunk 2 above)
Comment (paulmey) [inline integration-test/env/Extension/bin/_placeholder]: Makes sense, unnecessary for docker, but necessary for the humans... 😉

**Comment by ahmetalpbalkan** [comment]
@paulmey addressed CR comments, feel free to take a look again. Thanks for the openssl tips. I removed the checked-in certs.

I wrote the actual integration tests (it will be another PR) using the [bats](https://github.com/sstephenson/bats) framework and wrote a bunch of  I bash functions to deal with containerized tests and pushing certs/settings to the test container.

**Comment by paulmey** [comment]
Awesome! LGTM
Thanks, @ahmetalpbalkan!

**Comment by ahmetalpbalkan** [comment]
just added another commit

```
commit 9b7043a120c3dd99bb7a922b8c7b846ca9d90587

    - removed placeholder dirs (create them in dockerfile)
    - change handlermanifest to use the shim
```

**Comment by ahmetalpbalkan** [comment]
merging. @boumenot feel free to PTAL sometime

**Pull request merged**